[Serialization] Emit an error on deserializing an invalid declaration

bnbarham opened the pull request:
When running in the allow errors mode
(-experimental-allow-module-with-compiler-errors), modules may contain
invalid declarations. The rest of the compiler pipeline, however,
expects to have valid declarations unless diagnostics have emitted an
error. Emit an error while deserializing to maintain this assumption.

Note that these errors will not have a useful location, unless there's a
corresponding `.swiftsourceinfo`. This isn't a problem for the intended
use case in IDEs, where diagnostics outside the current file would be
ignored anyway.

Since reading declarations is lazy, SILGen (and thus SIL diagnostics)
can still run as long as any invalid declarations weren't referenced in
the compiling module.

Resolves rdar://74325388

bnbarham committed:
[Serialization] Emit an error on deserializing an invalid declaration

When running in the allow errors mode
(-experimental-allow-module-with-compiler-errors), modules may contain
invalid declarations. The rest of the compiler pipeline, however,
expects to have valid declarations unless diagnostics have emitted an
error. Emit an error while deserializing to maintain this assumption.

Note that these errors will not have a useful location, unless there's a
corresponding `.swiftsourceinfo`. This isn't a problem for the intended
use case in IDEs, where diagnostics outside the current file would be
ignored anyway.

Since reading declarations is lazy, SILGen (and thus SIL diagnostics)
can still run as long as any invalid declarations weren't referenced in
the compiling module.

Resolves rdar://74325388

bnbarham commented:
@swit-ci please test

bnbarham commented:
@swift-ci please test

xedin reviewed (commented):
Makes sense to me. I'd say it would be to add a check to invalid declaration diagnostic and just skip error emission if it's allowed to import invalid declarations.

bnbarham commented:
@xedin I've added a fatal diagnostic for the case when an invalid declaration was read in a module without allow errors enabled.

Also fixed up the test so it actually crashes in the case the radar mentions before the change (would still have failed, but not with the crash).

@swift-ci please test

xedin commented:
Looks good, thank you!
